Assets/BossRoom/Scripts/Shared/Infrastructure/PubSub/MessageChannel.cs

```diff
@@ -0,0 +1,64 @@
+using System;
+using System.Collections.Generic;
+using UnityEngine.Assertions;
+
+namespace BossRoom.Scripts.Shared.Infrastructure
+{
+    public class MessageChannel<T> : IMessageChannel<T>
+    {
+        private readonly List<Action<T>> m_MessageHandlers = new List<Action<T>>();
+        private readonly Queue<Action> m_PendingHandlers = new Queue<Action>();
+
+
+        public bool IsDisposed { get; private set; } = false;
+
+        public void Dispose()
+        {
+            if (!IsDisposed)
+            {
+                IsDisposed = true;
+                m_MessageHandlers.Clear();
+                m_PendingHandlers.Clear();
+            }
+        }
+
+        public virtual void Publish(T message)
+        {
+            while (m_PendingHandlers.Count > 0)
+            {
+                m_PendingHandlers.Dequeue()?.Invoke();
+            }
+
+            foreach (var messageHandler in m_MessageHandlers)
+            {
+                messageHandler?.Invoke(message);
+            }
+        }
+
+        public virtual IDisposable Subscribe(Action<T> handler)
+        {
+            Assert.IsTrue(!m_MessageHandlers.Contains(handler), "Attempting to subscribe with the same handler more than once");
+            m_PendingHandlers.Enqueue(() => { DoSubscribe(handler); });

```

Comment:
This queue of actions that would either add or remove subscriber is used to prevent problems from immediate modification of list of subscribers that's possible if you decide to unsubscribe in a message handler. 

I've added a piece of documentation in the field description of the queue itself.

Reply:
ah I see, so it's so we don't modify the list while iterating on it. With this, we'd unsubscribe during the next publish invocation basically correct?

Reply:
```suggestion
           // so we don't modify the handler list while iterating on it (which could happen if the handler unsubscribes). With this, we'd unsubscribe on next publish.
            m_PendingHandlers.Enqueue(() => { DoSubscribe(handler); });

```